File: exercism/c/binary-search/src/binary_search.c
#include "binary_search.h"
#include <stdio.h>

int *binary_search(int n, int array[], int length) {
  int a = 0, b = length - 1, mid, probe;

  while (a <= b) {
    mid = a + (b - a) / 2; // note: (a + b) / 2 could overflow
    probe = array[mid];
    if (probe == n)
      return &array[mid];
    if (probe < n) {
      a = mid + 1;
    } else {
      b = mid - 1;
    }
  }
  return (void *)0;
}

File: exercism/go/binary-search/binary_search.go
package binarysearch

func SearchInts(ints []int, x int) (i int) {
	for a, b := 0, len(ints); a < b; {
		mid := a + (b-a)>>1 // (a + b) / 2 can overflow
		probe := ints[mid]
		if x == probe {
			return mid
		}
		if x < probe {
			b = mid
		} else {
			a = mid + 1
		}
	}
	return -1
}
